File: examples/nodejs/license-client.mjs
/**
 * KISS license client untuk Node.js 18+, Bun, Deno.
 * Satu file, tanpa dependency. Pakai built-in `fetch`, `crypto`, `fs`.
 *
 * Cara pakai:
 *   import { LicenseClient } from "./license-client.mjs";
 *
 *   const client = new LicenseClient({
 *     apiUrl: "https://license.kin.my.id",
 *     signingKey: "<sama dengan SIGNING_KEY di .env VPS>",
 *     product: "myapp",
 *   });
 *
 *   // Pertama kali user input key:
 *   const res = await client.activate("VPXNC-YP98C-T4BH9-APW5Q");
 *   if (!res.valid) throw new Error(res.status);
 *
 *   // Setiap app start:
 *   if (!await client.check(savedKey)) process.exit(1);
 */
import crypto from "node:crypto";
import fs from "node:fs";
import os from "node:os";
import path from "node:path";

export class LicenseError extends Error {}

export class LicenseClient {
  constructor({
    apiUrl,
    signingKey,
    product,
    configDir,
    timeoutMs = 10_000,
    offlineGraceDays = 7,
  }) {
    this.apiUrl = apiUrl.replace(/\/$/, "");
    this.signingKey = signingKey;
    this.product = product;
    this.timeoutMs = timeoutMs;
    this.offlineGraceMs = offlineGraceDays * 86_400_000;
    this.configDir = configDir ?? path.join(os.homedir(), ".config", product);
    fs.mkdirSync(this.configDir, { recursive: true });
  }

  // ----- public -----
  activate(key, fingerprint = undefined) {
    return this._call("/v1/activate", {
      key, machine_id: this.machineId(), product: this.product, fingerprint,
    });
  }

  validate(key) {
    return this._call("/v1/validate", {
      key, machine_id: this.machineId(), product: this.product,
    });
  }

  deactivate(key) {
    return this._call("/v1/deactivate", {
      key, machine_id: this.machineId(), product: this.product,
    });
  }

  /**
   * Convenience: validate + grace period offline.
   * - Online & valid → true, simpan last_ok.
   * - Online & tidak valid → false.
   * - Offline / error jaringan → true kalau last_ok < N hari, else false.
   */
  async check(key) {
    try {
      const res = await this.validate(key);
      if (res.valid) {
        this._touchLastOk();
        return true;
      }
      return false;
    } catch (err) {
      if (err instanceof LicenseError) throw err;     // signature mismatch tetap fatal
      return this._withinGrace();
    }
  }

  machineId() {
    const p = path.join(this.configDir, "machine.id");
    if (!fs.existsSync(p)) fs.writeFileSync(p, crypto.randomUUID());
    return fs.readFileSync(p, "utf8").trim();
  }

  // ----- internal -----
  async _call(p, body) {
    const clean = Object.fromEntries(
      Object.entries(body).filter(([, v]) => v !== undefined && v !== null)
    );
    const ctl = new AbortController();
    const timer = setTimeout(() => ctl.abort(), this.timeoutMs);
    let data;
    try {
      const r = await fetch(this.apiUrl + p, {
        method: "POST",
        headers: { "content-type": "application/json" },
        body: JSON.stringify(clean),
        signal: ctl.signal,
      });
      if (!r.ok) throw new Error(`HTTP ${r.status}`);
      data = await r.json();
    } finally {
      clearTimeout(timer);
    }
    if (!this._verify({ ...data })) {
      throw new LicenseError("signature mismatch — koneksi tidak terpercaya");
    }
    return data;
  }

  _verify(resp) {
    const sig = resp.signature ?? "";
    delete resp.signature;
    const raw = JSON.stringify(sortKeys(resp));
    const expected = crypto.createHmac("sha256", this.signingKey)
      .update(raw).digest("hex");
    if (sig.length !== expected.length) return false;
    return crypto.timingSafeEqual(Buffer.from(sig), Buffer.from(expected));
  }

  _touchLastOk() {
    fs.writeFileSync(path.join(this.configDir, "last_ok"), String(Date.now()));
  }

  _withinGrace() {
    const p = path.join(this.configDir, "last_ok");
    if (!fs.existsSync(p)) return false;
    const last = parseInt(fs.readFileSync(p, "utf8").trim(), 10);
    if (!Number.isFinite(last)) return false;
    return (Date.now() - last) < this.offlineGraceMs;
  }
}

function sortKeys(o) {
  if (Array.isArray(o)) return o.map(sortKeys);
  if (o && typeof o === "object") {
    return Object.keys(o).sort().reduce((acc, k) => {
      acc[k] = sortKeys(o[k]);
      return acc;
    }, {});
  }
  return o;
}

// --- Quick CLI ---
if (import.meta.url === `file://${process.argv[1]}`) {
  const [, , action, key] = process.argv;
  if (!["activate", "validate", "deactivate", "check"].includes(action) || !key) {
    console.error("Usage: node license-client.mjs {activate|validate|deactivate|check} <KEY>");
    process.exit(1);
  }
  const client = new LicenseClient({
    apiUrl: process.env.LICENSE_API ?? "https://license.kin.my.id",
    signingKey: process.env.LICENSE_SIGNING_KEY ?? "",
    product: process.env.LICENSE_PRODUCT ?? "myapp",
  });
  if (!client.signingKey) {
    console.error("Set LICENSE_SIGNING_KEY env var (sama dengan SIGNING_KEY server).");
    process.exit(1);
  }
  const res = await client[action](key);
  console.log(typeof res === "boolean" ? (res ? "OK" : "BLOCKED") : JSON.stringify(res, null, 2));
}

File: examples/python/license_client.py
"""KISS license client for Python 3.9+.

Usage:
    from license_client import LicenseClient, LicenseError

    client = LicenseClient(
        api_url="https://license.kin.my.id",
        signing_key="<same as SIGNING_KEY in .env on the VPS>",
        product="myapp",
    )

    # First time the user enters a key:
    res = client.activate("VPXNC-YP98C-T4BH9-APW5Q")
    if not res["valid"]:
        raise LicenseError(res["status"])

    # On every app start:
    if not client.check(saved_key):
        sys.exit("Invalid license")

Only depends on `requests`. Install: pip install requests
"""
from __future__ import annotations

import hashlib
import hmac
import json
import os
import pathlib
import time
import uuid
from typing import Any

import requests


class LicenseError(RuntimeError):
    """Raised when the license is invalid or the response cannot be trusted."""


class LicenseClient:
    def __init__(
        self,
        api_url: str,
        signing_key: str,
        product: str,
        *,
        config_dir: pathlib.Path | None = None,
        timeout: float = 10.0,
        offline_grace_days: int = 7,
    ):
        self.api_url = api_url.rstrip("/")
        self.signing_key = signing_key.encode()
        self.product = product
        self.timeout = timeout
        self.offline_grace_seconds = offline_grace_days * 86400
        self.config_dir = config_dir or (pathlib.Path.home() / ".config" / product)
        self.config_dir.mkdir(parents=True, exist_ok=True)

    # ----- public -----
    def activate(self, key: str, fingerprint: str | None = None) -> dict[str, Any]:
        return self._call("/v1/activate", {
            "key": key,
            "machine_id": self.machine_id(),
            "product": self.product,
            "fingerprint": fingerprint,
        })

    def validate(self, key: str) -> dict[str, Any]:
        return self._call("/v1/validate", {
            "key": key,
            "machine_id": self.machine_id(),
            "product": self.product,
        })

    def deactivate(self, key: str) -> dict[str, Any]:
        return self._call("/v1/deactivate", {
            "key": key,
            "machine_id": self.machine_id(),
            "product": self.product,
        })

    def check(self, key: str) -> bool:
        """Convenience: validate + offline grace period.

        - Online & valid → True, store last_ok.
        - Online & invalid → False.
        - Offline / network error → True if last_ok < N days, else False.
        """
        try:
            res = self.validate(key)
        except requests.RequestException:
            return self._within_grace()

        if res.get("valid"):
            self._touch_last_ok()
            return True
        return False

    def machine_id(self) -> str:
        """Stable per-machine ID, stored in the config dir."""
        p = self.config_dir / "machine.id"
        if not p.exists():
            p.write_text(uuid.uuid4().hex)
        return p.read_text().strip()

    # ----- internal -----
    def _call(self, path: str, body: dict[str, Any]) -> dict[str, Any]:
        body = {k: v for k, v in body.items() if v is not None}
        r = requests.post(self.api_url + path, json=body, timeout=self.timeout)
        r.raise_for_status()
        data = r.json()
        if not self._verify(dict(data)):
            raise LicenseError("signature mismatch — untrusted response")
        return data

    def _verify(self, resp: dict[str, Any]) -> bool:
        sig = resp.pop("signature", "")
        raw = json.dumps(resp, sort_keys=True, separators=(",", ":")).encode()
        expected = hmac.new(self.signing_key, raw, hashlib.sha256).hexdigest()
        return hmac.compare_digest(sig, expected)

    def _touch_last_ok(self) -> None:
        (self.config_dir / "last_ok").write_text(str(int(time.time())))

    def _within_grace(self) -> bool:
        p = self.config_dir / "last_ok"
        if not p.exists():
            return False
        try:
            last = int(p.read_text().strip())
        except ValueError:
            return False
        return (int(time.time()) - last) < self.offline_grace_seconds


# --- Quick CLI for testing ---
if __name__ == "__main__":
    import argparse

    ap = argparse.ArgumentParser()
    ap.add_argument("action", choices=["activate", "validate", "deactivate", "check"])
    ap.add_argument("key")
    ap.add_argument("--api", default=os.getenv("LICENSE_API", "https://license.kin.my.id"))
    ap.add_argument("--key-secret", dest="signing_key",
                    default=os.getenv("LICENSE_SIGNING_KEY", ""))
    ap.add_argument("--product", default=os.getenv("LICENSE_PRODUCT", "myapp"))
    args = ap.parse_args()

    if not args.signing_key:
        raise SystemExit(
            "Set the LICENSE_SIGNING_KEY env var or use --key-secret. "
            "It must match the SIGNING_KEY in the server's .env file."
        )

    client = LicenseClient(args.api, args.signing_key, args.product)
    fn = getattr(client, args.action)
    res = fn(args.key)
    if isinstance(res, bool):
        print("OK" if res else "BLOCKED")
    else:
        print(json.dumps(res, indent=2))
